File: qt/multimedia/gen_qvideowidget.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QVideoWidget struct {
	h          uintptr
	isSubclass bool
}

// NewQVideoWidget constructs a new QVideoWidget object.
func NewQVideoWidget(parent *qt.QWidget) *QVideoWidget {
	g := newQVideoWidget(QVideoWidget_new((*QWidget)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

// NewQVideoWidget2 constructs a new QVideoWidget object.
func NewQVideoWidget2() *QVideoWidget {
	g := newQVideoWidget(QVideoWidget_new2())
	g.isSubclass = true
	return g
}

func (this *QVideoWidget) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QVideoWidget_MetaObject(this.h)))
}

func (this *QVideoWidget) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QVideoWidget_Metacast(this.h, param1_Cstring))
}

func QVideoWidget_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QVideoWidget_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QVideoWidget) VideoSink() *QVideoSink {
	return newQVideoSink(QVideoWidget_VideoSink(this.h))
}

func (this *QVideoWidget) AspectRatioMode() qt.AspectRatioMode {
	return (qt.AspectRatioMode)(QVideoWidget_AspectRatioMode(this.h))
}

func (this *QVideoWidget) SizeHint() *qt.QSize {
	_goptr := qt.UnsafeNewQSize(unsafe.Pointer(QVideoWidget_SizeHint(this.h)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QVideoWidget) SetFullScreen(fullScreen bool) {
	QVideoWidget_SetFullScreen(this.h, (bool)(fullScreen))
}

func (this *QVideoWidget) SetAspectRatioMode(mode qt.AspectRatioMode) {
	QVideoWidget_SetAspectRatioMode(this.h, (int)(mode))
}

func (this *QVideoWidget) FullScreenChanged(fullScreen bool) {
	QVideoWidget_FullScreenChanged(this.h, (bool)(fullScreen))
}

func (this *QVideoWidget) OnFullScreenChanged(slot func(fullScreen bool)) {
	QVideoWidget_connect_FullScreenChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoWidget_FullScreenChanged
func miqt_exec_callback_QVideoWidget_FullScreenChanged(cb intptr_t, fullScreen bool) {
	gofunc, ok := cgo.Handle(cb).Value().(func(fullScreen bool))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := (bool)(fullScreen)

	gofunc(slotval1)
}

func (this *QVideoWidget) AspectRatioModeChanged(mode qt.AspectRatioMode) {
	QVideoWidget_AspectRatioModeChanged(this.h, (int)(mode))
}

func (this *QVideoWidget) OnAspectRatioModeChanged(slot func(mode qt.AspectRatioMode)) {
	QVideoWidget_connect_AspectRatioModeChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoWidget_AspectRatioModeChanged
func miqt_exec_callback_QVideoWidget_AspectRatioModeChanged(cb intptr_t, mode int) {
	gofunc, ok := cgo.Handle(cb).Value().(func(mode qt.AspectRatioMode))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := (qt.AspectRatioMode)(mode)

	gofunc(slotval1)
}

func QVideoWidget_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QVideoWidget_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QVideoWidget_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QVideoWidget_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QVideoWidget) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QVideoWidget_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QVideoWidget) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QVideoWidget_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoWidget_MetaObject
func miqt_exec_callback_QVideoWidget_MetaObject(self QVideoWidget, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QVideoWidget{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QVideoWidget) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QVideoWidget_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QVideoWidget) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QVideoWidget_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QVideoWidget_Metacast
func miqt_exec_callback_QVideoWidget_Metacast(self QVideoWidget, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QVideoWidget{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/multimedia/gen_qvideowidget.cpp
// +build ignore

#include <QEvent>
#include <QHideEvent>
#include <QMetaObject>
#include <QMoveEvent>
#include <QObject>
#include <QPaintDevice>
#include <QResizeEvent>
#include <QShowEvent>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVideoSink>
#include <QVideoWidget>
#include <QWidget>
#include <qvideowidget.h>
#include "gen_qvideowidget.h"
class MiqtVirtualQVideoWidget : public virtual QVideoWidget {
public:
MiqtVirtualQVideoWidget(QWidget* parent): QVideoWidget(parent) {};
MiqtVirtualQVideoWidget(): QVideoWidget() {};

virtual ~MiqtVirtualQVideoWidget() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QVideoWidget::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QVideoWidget_MetaObject(const_cast<MiqtVirtualQVideoWidget*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QVideoWidget::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QVideoWidget::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QVideoWidget_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QVideoWidget::qt_metacast(param1);

	}
};
QVideoWidget* QVideoWidget_new(QWidget* parent) {
return new MiqtVirtualQVideoWidget(parent);
}
QVideoWidget* QVideoWidget_new2() {
return new MiqtVirtualQVideoWidget();
}
void QVideoWidget_virtbase(QVideoWidget* src
, QWidget** outptr_QWidget
) {
*outptr_QWidget = static_cast<QWidget*>(src);
}
QMetaObject* QVideoWidget_MetaObject(const QVideoWidget* self) {
	return (QMetaObject*) self->metaObject();
}
void* QVideoWidget_Metacast(QVideoWidget* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QVideoWidget_Tr(const char* s) {
	QString _ret = QVideoWidget::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
QVideoSink* QVideoWidget_VideoSink(const QVideoWidget* self) {
	return self->videoSink();
}
int QVideoWidget_AspectRatioMode(const QVideoWidget* self) {
	Qt::AspectRatioMode _ret = self->aspectRatioMode();
	return static_cast<int>(_ret);
}
QSize* QVideoWidget_SizeHint(const QVideoWidget* self) {
	return new QSize(self->sizeHint());
}
void QVideoWidget_SetFullScreen(QVideoWidget* self, bool fullScreen) {
	self->setFullScreen(fullScreen);
}
void QVideoWidget_SetAspectRatioMode(QVideoWidget* self, int mode) {
	self->setAspectRatioMode(static_cast<Qt::AspectRatioMode>(mode));
}
void QVideoWidget_FullScreenChanged(QVideoWidget* self, bool fullScreen) {
	self->fullScreenChanged(fullScreen);
}
void QVideoWidget_connect_FullScreenChanged(QVideoWidget* self, intptr_t slot) {
	MiqtVirtualQVideoWidget::connect(self, static_cast<void (QVideoWidget::*)(bool)>(&QVideoWidget::fullScreenChanged), self, [=](bool fullScreen) {
		bool sigval1 = fullScreen;
		miqt_exec_callback_QVideoWidget_FullScreenChanged(slot, sigval1);
	});
}
void QVideoWidget_AspectRatioModeChanged(QVideoWidget* self, int mode) {
	self->aspectRatioModeChanged(static_cast<Qt::AspectRatioMode>(mode));
}
void QVideoWidget_connect_AspectRatioModeChanged(QVideoWidget* self, intptr_t slot) {
	MiqtVirtualQVideoWidget::connect(self, static_cast<void (QVideoWidget::*)(Qt::AspectRatioMode)>(&QVideoWidget::aspectRatioModeChanged), self, [=](Qt::AspectRatioMode mode) {
		Qt::AspectRatioMode mode_ret = mode;
		int sigval1 = static_cast<int>(mode_ret);
		miqt_exec_callback_QVideoWidget_AspectRatioModeChanged(slot, sigval1);
	});
}
struct miqt_string QVideoWidget_Tr2(const char* s, const char* c) {
	QString _ret = QVideoWidget::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QVideoWidget_Tr3(const char* s, const char* c, int n) {
	QString _ret = QVideoWidget::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QVideoWidget_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQVideoWidget*>( (QVideoWidget*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QVideoWidget_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQVideoWidget*)(self) )->virtualbase_MetaObject();
}
void QVideoWidget_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQVideoWidget*>( (QVideoWidget*)(self) )->handle__Metacast = slot;
}
void* QVideoWidget_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQVideoWidget*)(self) )->virtualbase_Metacast(param1);
}
void QVideoWidget_Delete(QVideoWidget* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQVideoWidget*>( self );
	} else {
		delete self;
	}
}
